man/cbsd.8
.Dd December 10, 2016
.Dt cbsd 8
.Os
.Sh NAME
.Nm cbsd
.Nd FreeBSD jail management tool
.Sh SYNOPSIS
.Nm cbsd
.Op cmd
.Nm cbsd
.Op cmd
.Op param=val
.Nm cbsd
.Op --help
.Sh DESCRIPTION
.Nm
is a jail and VM management tool. It builds upon and extends standard
tools used for jails and bhyve manipulation.
.Pp
List of some of the programs and functions involved in
.Nm
:
.Bl -dash -width Ds
.It
vnet (VIMAGE)
.It
zfs
.It
racct/rctl
.It
pf/ipfw/ipfilter
.It
carp
.It
hastd
.It
bhyve
.It
Xen
.El

Many of these subsystems are not directly related to jails, however allow
.Nm
(being the connecting link between these components) to provide the system
administrator with a more advanced and integrated system for problem solving.
.Pp
The information provided below could be useful to the system administrator
and describes the
.Nm
architecture in more detail. It is also recommended to read through the
official articles about FreeBSD Jails for better understanding of the jails
basics.
.Pp
You can get a description for each command via:
.Nm Cm <command> --help
or
.Nm Cm help <command>
, for example, start with:

.Dl % cbsd help --help

.Ss Naming conventions
Let us agree on some names and values that will be used
throughout the documentation:

.Bl -tag -width Ds
.It node
physical server / resource unit
.It jail
isolated environment with its own set of software/services.
.It farm
cluster of interconnected nodes, a full-fledged peer network (each node can do other tasks)
.It basejail
in the CBSD context - a copy of FreeBSD base system.
.It cbsd
an entity that has control over the particular node(s) and certain
subsystems of FreeBSD, is able to provide a simplified and unified action
(API) on the nodes or jails and provides ACL and setting permissions for
CBSD customers
.It $jname
jail name

All working data of CBSD is placed in the $workdir directory (e.g. /usr/jails),
it is also the home directory for the cbsd user. You can quickly change to this directory by issuing:

.Dl % cd ~cbsd

The most important data are located in $workdir/jails-data/$jname,
as it is directly the root of the jail's filesystem called $jname,
if jail is created with the flag baserw=1.
Or contains the data of jail, which are superimposed on the standard $workdir/basejail/$basename
system mounted in the ${workdir}/jails directory.

The base system can always be built from source code,
or downloaded from the repository.

.Nm
allows jails creation, backups, snapshotting, cloning and changing different
jails options. It also allows to control the order in which the jails should be
started. When using more than one node running
.Nm
it allows to clone jails between nodes and even allows to
control the remote node(s) from the "master" of the farm.
.Pp
All commands begin with
.Nm
prefix.
.Pp
The
.Nm
commands can accept arguments which are written in a "param=value" form.
.Nm
commands can have both obligatory arguments and arguments which are optional.
To receive list of all available commands type: cbsd --help
.Pp
Each subcommand has a short usage help available when typing cbsd <subcommand>
--help and an extended manual page accessible by typing "man cbsd-subcommand".
The subcommands are as follows:
.Bl -tag -width Ds
.It Cm alljails Ns = Ns Ar 1

.Sh ENVIRONMENT
By default, CBSD displays text in color (using escape sequences ANSII).
If for some reason the color is unpleasant, or you want to use output from
CBSD utilities in your scripts, you can disable the color through the
environment variable NOCOLOR=1. For example:

.Dl % env NOCOLOR=1 cbsd jls

In addition, you can create your own COLOR schemes via ~cbsd/etc/color.conf 
config file. For example for white color schema you can copy sample profile:

.Bd -literal -offset indent
% cp /usr/local/cbsd/etc/defaults/color-white.conf \\
  ~cbsd/etc/color.conf
.Ed
.Pp

.Sh FILES
.Bl -tag -width Ds -compact
.It ${workdir}
Working directory on CBSD node (usually /usr/jails).
It is initialized through
.Nm Cm initenv
.It ${workdir}/.rssh/
Directory to store private keys of remote nodes. The files are added and
removed via the
.Nm Cm node
.It ${workdir}/.ssh
This location serves as a storage for private and a public keys of the node itself.
Formed during the initialization with the
.Nm Cm initenv
command. The public keys of remote hosts when adding them as node using the
.Nm Cm node mode=add
will be taken from here. Key filename is the MD5 sum of the node name.
.It ${workdir}/basejail
Here are the FreeBSD base system and kernels kept ready for use (the result of
.Nm Cm buildworld/buildkernel,
.Nm Cm installworld/installkernel
or
.Nm Cm repo action=get sources=base/kernel
)
.It ${workdir}/etc
Configuration files needed to run CBSD
.It ${workdir}/export
Default directory for jails exported to a file (using the
.Nm Cm jexport jname=$jname,
the file will be stored under $jname.img)
.It ${workdir}/import
Default directory from which jails are imported (using
.Nm Cm jimport jname=$jname,
jail will be deployed as $jname)
.It ${workdir}/jails
This directory contains the mount point for the jails' root, those that use baserw=0.
.It ${workdir}/jails-data
This directory has the jails data. These are the directories that you need to backup (including fstab and rc.conf files).
Also, if the jail uses baserw=1, these directories contain the jail's root when it starts
.It ${workdir}/jails-fstab
fstab file for the jails. The syntax is like for regular FreeBSD fstab with the only exception,
that the path to the mount point is written relative to the jail's root
(record
.Dl /usr/ports /usr/ports nullfs rw 0 0
in the file fstab.$jname means,
that the master node directory /usr/ports will be mounted at startup in ${workdir}/jails/$jname/usr/ports)
.It ${workdir}/jails-rcconf
rc.conf files for jail creation. These parameters can be changed with your editor,
or via the command
.Nm Cm jset $jname param=val
(eg cbsd jset jname=$jname ip="192.168.0.2/24"). To change these settings the jail should be turned off.
.It ${workdir}/jails-system
This directory may contain some helper scripts related to the jail
(e.g. configuration wizards, etc) as well as the preserved jail traffic information,
when using ipfw and its description. This directory participates in jimport/jexport
operations and migration of jail
.It ${workdir}/var
directory that contains system information for CBSD. For example, in ${workdir}/var/db
is an inventory of local and remote nodes that were added
.It /usr/local/cbsd
A copy of the original files installed by CBSD port. Also contains the working scripts in sudoexec

.Sh EXIT STATUS
.Ex -std
The codes are usually described in the --help section for the command/subcommand.

.Sh EXAMPLES
.Tp
Show list of jails:
.Pp
.Nm Cm jls
.Pp
.Tp
Show help for jlogin command:
.Pp
.Nm Cm jlogin Fl Fl help
.Pp
.Tp
Run DIALOG-based jail configuration tools:
.Pp
.Nm Cm jconstruct-tui
.Pp
.Sh EXIT STATUS
cbsd returns a zero exit status if it succeeds. Non-zero is returned in case
of a failure or non-standard conclusion.
The codes are usually described in the --help parameter to a command/subcommand.
.Sh SEE ALSO
.Xr jail 8

.Sh AUTHORS
.An Oleg Ginzburg Aq Mt olevole@olevole.ru


tools/history
#!/usr/local/bin/cbsd
#v10.0.3
MYARG=""
MYOPTARG="line"
MYDESC="Show cbsd history command"
ADDHELP="line=N for output only last N commands\n"
EXTHELP="wf_history"
CBSDMODULE="sys"

. ${subr}

line=

. ${cbsdinit}

TAIL_OPT=

[ -n "${line}" ] && TAIL_OPT="-n${line}"
[ -r "${workdir}/.cbsd_history" ] && ${TAIL_CMD} ${TAIL_OPT} ${workdir}/.cbsd_history


cbsd.lua
product="CBSD"
myversion="11.2.1d"

if not workdir then
	print ( "no workdir" )
	os.exit(1)
end

distdir="/usr/local/cbsd"
subr= workdir .. "/nc.lua"
nodenamefile= workdir .. "/nodename"
settingstui= workdir .. "/settings-tui.subr"
dialog= workdir .. "/dialog.subr"
tools= workdir .. "/tools.subr"
mdtools= workdir .. "/mdtools.subr"
zfstool= workdir .. "/zfs.subr"
jfs= workdir .. "/jfs.subr"
color= workdir .. "/ansiicolor.subr"
nodes= workdir .. "/nodes.subr"
jrcconf= workdir .. "/rcconf.subr"
vimageconf= workdir .. "/vnet.subr"
vimagetui= workdir .. "/vnet-tui.subr"
inventory= workdir .. "/nc.inventory"
nodedescr= workdir .. "/node.desc"
initenv= workdir .. "/initenv.subr"
system= workdir .. "/system.subr"
buildconf= workdir .. "/build.subr"
distccacheconf= workdir .. "/distccache.subr"
base2pkgconf= workdir .. "/base2pkg.subr"
mailconf= workdir .. "/mailtools.subr"
strings= workdir .. "/strings.subr"
miscdir= workdir .. "/misc"
jobdir= workdir .. "/job"
srcdir= workdir .. "/src"
tmpdir= workdir .. "/tmp"
ftmpdir= workdir .. "/ftmp"
importdir= workdir .. "/import"
exportdir= workdir .. "/export"
basejaildir= workdir .. "/basejail"
basejailpref="base"
jaildatadir= workdir .. "/jails-data"
jailfstabdir= workdir .. "/jails-fstab"
jailrcconfdir= workdir .. "/jails-rcconf"
jailfstabpref="fstab."
jaildir= workdir .. "/jails"
jaildatapref="data"
jailsysdir= workdir .. "/jails-system"
bindir= workdir .. "/bin"
etcdir= workdir .. "/etc"
jailctldir= workdir .. "/jailctl"
nodectl= workdir .. "/nodectl"
rcddir= workdir .. "/rc.d"
sbindir= workdir .. "/sbin"
systemdir= workdir .. "/system"
moduledir= workdir .. "/modules"
toolsdir= workdir .. "/tools"
upgradedir= workdir .. "/upgrade"
vardir= workdir .. "/var"
spooldir= vardir .. "/spool"
taskdir= spooldir .. "/task"
rundir= vardir .. "/run"
maildir= vardir .. "/mail"
logdir= vardir .. "/log"
sharedir= workdir .. "/share"
dbdir= vardir .. "/db"
localcbsdconffile= "/cbsd.lua"
localcbsdconf= workdir .. localcbsdconffile
sudoexecdir= workdir .. "/sudoexec"
systemsudoexecdir= distdir .. "/sudoexec"
cbsduser="cbsd"
production="0"
rsshdir= workdir .. "/.rssh"
sshdir= workdir .. "/.ssh"
templateldir= workdir .. "/template"
fwcount_st="99"
fwcount_end="2000"
greeting="node"
jailmapdb= dbdir .. "/jmap.txt"
-- external source for online doc
docversion="10.3"
cbsddocsrc="https://www.bsdstore.ru/en" .. docversion .. "/"

f=io.open(nodenamefile,"r")
if f ~= nil then
	nodename = f:read("*l")
	io.close(f)
end

-- if [ -n "${NOCOLOR}" ]; then
--    ECHO="echo"
-- else
--    ECHO="echo -e"
--    [ -f "${color}" ] && . $color
-- fi

